### PR TITLE
fix issue from #1017

### PR DESCRIPTION
**Pull request by matthieugomez**

### CODE REVIEW

File: test/formula.jl [hunk 1]
```diff
@@ -535,4 +535,11 @@ df = DataFrame(x = [1,2,3], y = [4,5,6])
 mf = ModelFrame(y ~ 1 + (1 | x), df)
 @test coefnames(mf) == ["(Intercept)"]
 
+
+# Ensure X is not a view on df column
+df = DataFrame(x = [1.0,2.0,3.0], y = [4.0,5.0,6.0])
+mf = ModelFrame(y ~ 0 + x, df)
+X = ModelMatrix(mf).m
+X[1] = 0.0
+@assert isapprox(mf.df[1, :x], 1.0)
```
Comment (ararslan) [inline test/formula.jl]: Should be `@test` rather than `@assert`. Also this test is causing the failure on all platforms.

**Comment by nalimilan** [comment]
Actually, I don't think using `reduce` here was a good idea at all: it's going to create as many temporary copies as the number of columns, minus 1. What we should do is allocate the matrix, then fill it with a loop (`hcat(blocks...)` would work, but splatting isn't recommended with a variable number of arguments).

**Comment by matthieugomez** [comment]
I don't understand the code so I'm not going to do that.

**Comment by nalimilan** [comment]
You don't need to understand the code to do that (and actually I don't really understand it either).

**Comment by matthieugomez** [comment]
How to compute the number of columns ex ante?

**Comment by matthieugomez** [comment]
Oh. Ok I'll just do a loop afterwards.

File: src/statsmodels/formula.jl [hunk 2]
```diff
@@ -481,7 +481,16 @@ creating the model matrix.
...
 
-    ModelMatrix{T}(reduce(hcat, blocks), assign)
+    I = size(dfrm, 1)
+    J = mapreduce(x -> size(x, 2), +, blocks)
+    X = zeros(eltype(T), I, J)
+    i = 1
+    for block in blocks
+        len = size(block, 2)
+        X[:, i:(i + len - 1)] = block
+        i += len
```
Comment (ararslan) [inline src/statsmodels/formula.jl]: What's this `i` doing? Doesn't appear to be used after the loop.

Comment (matthieugomez) [inline src/statsmodels/formula.jl]: it keeps track of the current column in X. What do you suggest?

Comment (ararslan) [inline src/statsmodels/formula.jl]: WOW, I'm dumb! I don't know how I missed it being used literally in the line above where I commented. Sorry about that...

File: src/statsmodels/formula.jl [hunk 3]
```diff
@@ -481,7 +481,16 @@ creating the model matrix.
         append!(assign, fill(i_term, size(blocks[end], 2)))
     end
 
-    ModelMatrix{T}(reduce(hcat, blocks), assign)
+    I = size(dfrm, 1)
+    J = mapreduce(x -> size(x, 2), +, blocks)
+    X = zeros(eltype(T), I, J)
```
Comment (nalimilan) [inline src/statsmodels/formula.jl]: @GordStephen Is this OK as regards your recent work on sparse matrices? Or should we use `similar` here?

File: test/formula.jl [hunk 4]
```diff
@@ -535,4 +535,11 @@ df = DataFrame(x = [1,2,3], y = [4,5,6])
 mf = ModelFrame(y ~ 1 + (1 | x), df)
 @test coefnames(mf) == ["(Intercept)"]
 
+
+# Ensure X is not a view on df column
+df = DataFrame(x = [1.0,2.0,3.0], y = [4.0,5.0,6.0])
+mf = ModelFrame(y ~ 0 + x, df)
+X = ModelMatrix(mf).m
+X[1] = 0.0
+@test isapprox(mf.df[1, :x], 1.0)
```
Comment (nalimilan) [inline test/formula.jl]: Do you really need `isapprox` instead of `==` here? (Note you can also `≈` instead.)

**Comment by matthieugomez** [comment]
I'm sorry but dont have time to work more on that. Feel free to use to use the code in another pull request.

**Comment by nalimilan** [comment]
See https://github.com/JuliaStats/DataFrames.jl/pull/1048.

**Comment by matthieugomez** [comment]
Thanks

**Pull request closed without merging**